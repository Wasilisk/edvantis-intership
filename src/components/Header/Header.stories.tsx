import React from 'react';
import {ComponentMeta, Story} from '@storybook/react';
import Header from './index';
import {HeaderType} from "../../models";

const mockHeader: HeaderType = {
    logo: {
        url: "https://res2.weblium.site/res/5e2f1da2c947890021c3f194/5e33277186cda400222c9970_optimized_259?nowebp"
    },
    navigation: [
        {
            name: "Promotions",
            to: "/link"
        },
        {
            name: "Best-selling",
            to: "/link"
        },
        {
            name: "Collection",
            to: "/link"
        },
        {
            name: "Testimonials",
            to: "/link"
        },
        {
            name: "Blog",
            to: "/link"
        },
        {
            name: "Contact us",
            to: "/link"
        }
    ],
};

export default {
    title: 'Application/Header',
    component: Header,
} as ComponentMeta<typeof Header>;

const Template: Story<HeaderType> = (props) => <Header {...props}/>;

export const Default = Template.bind({});
Default.args = {
    ...mockHeader,
}


