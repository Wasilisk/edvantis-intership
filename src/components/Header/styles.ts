import styled from "styled-components";

export const HeaderStyles = styled.div`
  max-width: 1200px;
  padding: 30px 40px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  box-sizing: border-box;
  margin: 0 auto;
  
  .logo-block {
    flex-grow: 1;
    display: flex;
    align-items: center;
    z-index: 2;
    
    img {
      height: 32px;
      width: auto;
    }
  }
  
  & > *:not(:last-child) {
    margin-right: 10px;
  }

  @media (max-width: 768px) {
    padding: 20px 20px;
  }
`