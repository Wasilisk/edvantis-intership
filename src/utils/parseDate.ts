const month = ["January","February","March","April","May","June","July","August","September","October","November","December"];

export const parseDate = (timestamp: number) => {
    const date = new Date(timestamp);

    return `${month[date.getMonth()]} ${('0' + date.getDate()).slice(-2)}, ${date.getFullYear()}`;
}