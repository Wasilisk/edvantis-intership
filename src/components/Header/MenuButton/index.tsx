import React from 'react';
import {MenuButtonStyles } from './styles';
import {MenuButtonProps} from "./types";

const MenuButton = ({isOpen, setIsOpen}: MenuButtonProps) => {

    const onClick = () => setIsOpen(!isOpen);

    return (
        <MenuButtonStyles isOpen={isOpen} onClick={onClick}>
            <span/>
            <span/>
            <span/>
        </MenuButtonStyles>
    );
};

export default MenuButton;