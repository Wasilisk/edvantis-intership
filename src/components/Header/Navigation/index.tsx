import React from 'react';
import {NavigationStyles} from "./styles";
import {NavigationProps} from "./types";
import {Link} from "react-router-dom";

const Navigation = ({links, isLoading}: NavigationProps) => {
    return (
        <NavigationStyles>
            {
                !isLoading && links && links.map((link, index) => <li
                        key={index}
                    >
                        <Link to={link.to}>{link.name}</Link>
                    </li>)
            }
        </NavigationStyles>
    );
};

export default Navigation;