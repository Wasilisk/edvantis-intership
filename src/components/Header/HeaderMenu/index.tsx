import React, {useState} from 'react';
import MenuIcon from "../MenuButton";
import Navigation from "../Navigation";
import Button from "../../common/Button";
import { HeaderMenuStyles } from './styles';
import {HeaderMenuProps} from "./types";
import {useDisableBodyScroll} from "../../../hooks/useDisableBodyScroll";

const HeaderMenu = ({links, isLoading}: HeaderMenuProps) => {
    const [isOpen, setIsOpen] = useState<boolean>(false);
    useDisableBodyScroll(isOpen);

    return (
        <>
            <MenuIcon isOpen={isOpen} setIsOpen={setIsOpen}/>
            <HeaderMenuStyles isOpen={isOpen} onClick={() => setIsOpen(false)}>
                <Navigation isLoading={isLoading} links={links}/>
                <Button variant="outlined" size="small" label="Shop Now" onClick={() => {}}/> //ToDo add OnClick logic
            </HeaderMenuStyles>
        </>
    );
};

export default HeaderMenu;