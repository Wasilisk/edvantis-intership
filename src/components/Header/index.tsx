import React from 'react';
import {HeaderStyles} from "./styles";
import Navigation from "./Navigation";
import Button from "../common/Button";
import {useAppSelector, useMediaQuery} from "../../hooks";
import HeaderMenu from "./HeaderMenu";
import {selectHeader, selectIsLayoutLoading} from "../../store/slices/AppLayout/selectors";
import Skeleton from "../common/Skeleton";

const Header = () => {
    const isMobile = useMediaQuery("(max-width: 1024px)");
    const headerInfo = useAppSelector(selectHeader);
    const isHeaderLoading = useAppSelector(selectIsLayoutLoading);

    return (
        <HeaderStyles>
            <div className="logo-block">
                {
                    isHeaderLoading
                        ? <Skeleton variant="rounded" height="34px" width="110px"/>
                        : headerInfo && <img src={headerInfo.logo.url} alt="Logo"/>
                }
            </div>
            {
                isMobile
                    ? <HeaderMenu isLoading={isHeaderLoading} links={headerInfo && headerInfo.navigation}/>
                    : <>
                        <Navigation isLoading={isHeaderLoading} links={headerInfo && headerInfo.navigation}/>
                        <Button variant="outlined" size="small" label="Shop Now"/>
                    </>
            }
        </HeaderStyles>
    );
};

export default Header;