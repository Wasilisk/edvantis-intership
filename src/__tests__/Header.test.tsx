import Header from "../components/Header";
import {render} from "./utils/store-wrapper";
import {mockHeader} from "./__mock__";

describe('Header component', () => {
    it('should render header component', () => {
        const {container, asFragment, getAllByAltText} = render(<Header/>, {
            initialState: {
                app_layout: {
                    header: mockHeader,
                    isLoading: false
                }
            }
        });
        expect(container).toHaveTextContent('Shop Now');
        expect(container).toHaveTextContent(mockHeader.navigation[1].name);
        expect(getAllByAltText("Logo")).toBeDefined();
        expect(asFragment()).toMatchSnapshot();
    })
    it('should render header component while header data is loading', () => {
        const {container, asFragment} = render(<Header/>, {
            initialState: {
                app_layout: {
                    isLoading: true
                }
            }
        });
        expect(container).toHaveTextContent('Shop Now');
        expect(container).not.toHaveTextContent(mockHeader.navigation[2].name);
        expect(container).not.toHaveTextContent(mockHeader.logo.url);
        expect(asFragment()).toMatchSnapshot();
    })
})