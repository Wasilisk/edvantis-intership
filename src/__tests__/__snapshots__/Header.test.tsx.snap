// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render header component 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ djOdHE"
  >
    <div
      class="logo-block"
    >
      <img
        alt="Logo"
        src="logo.jpg"
      />
    </div>
    <ul
      class="sc-gsnTZi gigZEl"
    >
      <li>
        <a
          href="/link1"
        >
          Link 1
        </a>
      </li>
      <li>
        <a
          href="/link2"
        >
          Link 2
        </a>
      </li>
      <li>
        <a
          href="/link3"
        >
          Link 3
        </a>
      </li>
      <li>
        <a
          href="/link4"
        >
          Link 4
        </a>
      </li>
    </ul>
    <button
      class="sc-dkzDqf jEwdBc"
    >
      Shop Now
    </button>
  </div>
</DocumentFragment>
`;

exports[`Header component should render header component while header data is loading 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ djOdHE"
  >
    <div
      class="logo-block"
    >
      <div
        class="sc-jSMfEi jgEOfq"
        height="34px"
        width="110px"
      />
    </div>
    <ul
      class="sc-gsnTZi gigZEl"
    />
    <button
      class="sc-dkzDqf jEwdBc"
    >
      Shop Now
    </button>
  </div>
</DocumentFragment>
`;
